File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/BackslashShouldBeAvoidedInAspNetRoutes.AspNet4x.cs
using System.Web.Mvc;
using System;

[Route(@"A\[controller]")]    // Noncompliant {{Replace '\' with '/'.}}
//     ^^^^^^^^^^^^^^^^^
public class BackslashOnControllerUsingVerbatimStringController : Controller { }

[Route("A\\[controller]")]    // Noncompliant {{Replace '\' with '/'.}}
//     ^^^^^^^^^^^^^^^^^
public class BackslashOnControllerUsingEscapeCharacterController : Controller { }

[Route("A\\[controller]\\B")] // Noncompliant {{Replace '\' with '/'.}}
//     ^^^^^^^^^^^^^^^^^^^^
public class MultipleBackslashesOnController : Controller { }

public class BackslashOnActionUsingVerbatimStringController : Controller
{
    [Route(@"A\[action]")]    // Noncompliant {{Replace '\' with '/'.}}
    //     ^^^^^^^^^^^^^
    public ActionResult Index() => View();
}

public class BackslashOnActionUsingEscapeCharacterController : Controller
{
    [Route("A\\[action]")]    // Noncompliant
    //     ^^^^^^^^^^^^^
    public ActionResult Index() => View();
}

public class MultipleBackslashesOnActionController : Controller
{
    [Route("A\\[action]\\B")] // Noncompliant
    //     ^^^^^^^^^^^^^^^^
    public ActionResult Index() => View();
}

[Route("\\[controller]")]    // Noncompliant
//     ^^^^^^^^^^^^^^^^
public class RouteOnControllerStartingWithBackslashController : Controller { }

public class AController : Controller
{
    //[Route("A\\[action]")]  // Compliant: commented out
    public ActionResult WithoutRouteAttribute() => View();

    [Route("A\\[action]", Name = "a", Order = 3)] // Noncompliant
    public ActionResult WithOptionalAttributeParameters() => View();

    [RouteAttribute("A\\[action]")] // Noncompliant
    public ActionResult WithAttributeSuffix() => View();

    [System.Web.Mvc.RouteAttribute("A\\[action]")] // Noncompliant
    public ActionResult WithFullQualifiedName() => View();

    [Route("A\\[action]")]  // Noncompliant
    [Route("B\\[action]")]  // Noncompliant
    [Route("C/[action]")]   // Compliant: forward slash is used
    public ActionResult WithMultipleRoutes() => View();

    [Route("A%5C[action]")] // Compliant: URL-escaped backslash is used
    public ActionResult WithUrlEscapedBackslash() => View();

    [Route("A/{s:regex(^(?!index\\b)[[a-zA-Z0-9-]]+$)}.html")]                 // Compliant: backslash is in regex
    public ActionResult WithRegexContainingBackslashInLookahead()  => View();

    [Route("A/{s:datetime:regex(\\d{{4}}-\\d{{2}}-\\d{{4}})}/B")]              // Compliant: backslash is in regex
    public ActionResult WithRegexContainingBackslashInMetaEscape() => View();

    [Route("A\\{s:regex(.*)\\[action]")]                                       // Noncompliant
    public ActionResult WithFirstBackslashBeforeFirstRegex() => View();

    [Route("{s:regex(.*)\\[action]")]                                          // FN: the backslash is not in the regex
    public ActionResult WithFirstBackslashAfterFirstRegex() => View();

    [Route("{s:regex([^\\\\]*)\\[action]")]                                    // FN: the second backslash is not in the regex
    public ActionResult WithSecondBackslashNotInRegex() => View();

    [Route("{s:regex([^\\\\]*)/regex([^\\\\]*)")]                              // Compliant: both backslashes are in the regex
    public ActionResult WithTwoBackslashesInRegex() => View();
}

namespace WithAliases
{
    using MyRoute = RouteAttribute;
    using ASP = System.Web;

    [MyRoute(@"A\[controller]")]            // Noncompliant
    public class WithAliasedRouteAttributeController : Controller { }

    [ASP.Mvc.RouteAttribute("A\\[action]")] // Noncompliant
    public class WithFullQualifiedPartiallyAliasedNameController : Controller { }
}

namespace WithFakeRouteAttribute
{
    [Route(@"A\[controller]")]      // Compliant: not a real RouteAttribute
    public class AControllerController : Controller { }

    [AttributeUsage(AttributeTargets.Class)]
    public class RouteAttribute : Attribute
    {
        public RouteAttribute(string template) { }
    }
}

class WithTuples
{
    void Test()
    {
        var tuple = (1, "A\\[controller]"); // Compliant: the argument is not in a method invocation
    }
}

// https://github.com/SonarSource/sonar-dotnet/issues/9193
namespace AttributeWithNamedArgument
{
    [AttributeUsage(AttributeTargets.All)]
    public class MyAttribute : Attribute
    {
        public string Name { get; set; }
    }

    public class MyController : Controller
    {
        [MyAttribute(Name = "Display HR\\Recruitment report")]
        public const string Text = "ABC";
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/BackslashShouldBeAvoidedInAspNetRoutes.AspNet4x.vb
Imports System
Imports System.Web.Mvc

<Route("A\[controller]")>   ' Noncompliant ^8#16 {{Replace '\' with '/'.}}
Public Class BackslashOnController
    Inherits Controller
End Class

<Route("A\[controller]\B")> ' Noncompliant ^8#18 {{Replace '\' with '/'.}}
Public Class MultipleBackslashesOnController
    Inherits Controller
End Class

Public Class BackslashOnActionController
    Inherits Controller

    <Route("A\[action]")>   ' Noncompliant ^12#12 {{Replace '\' with '/'.}}
    Public Function Index() As ActionResult
        Return View()
    End Function
End Class

Public Class MultipleBackslashesOnActionController
    Inherits Controller

    <Route("A\[action]\B")> ' Noncompliant ^12#14 {{Replace '\' with '/'.}}
    Public Function Index() As ActionResult
        Return View()
    End Function
End Class

<Route("\[controller]")>    ' Noncompliant
Public Class RouteOnControllerStartingWithBackslashController
    Inherits Controller
End Class

Public Class AController
    Inherits Controller

    Public Function WithoutRouteAttribute() As ActionResult             ' Compliant
        Return View()
    End Function

    <Route("A\[action]", Name:="a", Order:=3)>                          ' Noncompliant
    Public Function WithOptionalAttributeParameters() As ActionResult
        Return View()
    End Function

    <RouteAttribute("A\[action]")>                                      ' Noncompliant
    Public Function WithAttributeSuffix() As ActionResult
        Return View()
    End Function

    <System.Web.Mvc.RouteAttribute("A\[action]")>                       ' Noncompliant
    Public Function WithFullQualifiedName() As ActionResult
        Return View()
    End Function

    <Route("A\[action]")>                                               ' Noncompliant
    <Route("B\[action]")>                                               ' Noncompliant
    <Route("C/[action]")>                                               ' Compliant: forward slash is used
    Public Function WithMultipleRoutes() As ActionResult
        Return View()
    End Function

    <Route("A%5C[action]")>                                             ' Compliant: URL-escaped backslash is used
    Public Function WithUrlEscapedBackslash() As ActionResult
        Return View()
    End Function

    <Route("A/{s:regex(^(?!index\b)[[a-zA-Z0-9-]]+$)}.html")>           ' Compliant: backslash is in regex
    Public Function WithRegexContainingBackslashInLookahead() As ActionResult
        Return View()
    End Function

    <Route("A/{s:datetime:regex(\d{{4}}-\d{{2}}-\d{{4}})}/B")>          ' Compliant: backslash is in regex
    Public Function WithRegexContainingBackslashInMetaEscape() As ActionResult
        Return View()
    End Function
End Class

' https://github.com/SonarSource/sonar-dotnet/issues/9193
Namespace AttributeWithNamedArgument
    <AttributeUsage(AttributeTargets.All)>
    Public Class MyAttribute
        Inherits Attribute
        Public Property Name As String
    End Class

    Public Class MyController
        Inherits Controller
        <MyAttribute(Name:="Display HR\Recruitment report")>
        Public Const Text As String = "ABC"
    End Class
End Namespace
